Allow search indexing to be disabled

jamos-tay opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Enhancement to an existing feature

<!--
    If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"
-->

Fixes #514 

<!--
    Please ensure your pull request is ready:
    - Bug fix PR that is non-trivial **should** add a page or unit test for regression testing.
    - Feature PR **must** add a page to the user guide for demo.
    - Enhancement PR **should** update the user guide.

    Otherwise, prefix your PR title with "[WIP]".
-->

**What is the rationale for this request?**

Allows the user to disable indexing of headings for search.

**What changes did you make? (Give an overview)**

Added a `disableSearch` option in the `site.json`

If set to `true`, all pages will be marked as unsearchable, so they won't be indexed in sitedata.

`setupWithSearch` is still called, but it does nothing.

**Testing instructions:**

Not sure how to test this...

yamgent commented:
> `setupWithSearch` is still called, but it does nothing.

Ok, my apologies, maybe I wasn't clear with the original issue, but the essential part of it is that `setupWithSearch()` **must not be called.**

This is because `setupWithSearch()` will load the `searchbar` component. The `searchbar` component may interfere with some search functionalities provided by other search providers. Hence it must not be loaded, and hence only `setup()` should be called.

jamos-tay commented:
@yamgent Sure, no problem, added a check

Revamped docs and rebased to fix merge conflict.

Also, this might be unrelated, but did the recent doc revamp cause any issues? I wasn't able to deploy the docs, getting these errors:

![image](https://user-images.githubusercontent.com/19278089/50681466-8e242780-1046-11e9-884a-b473ca066d54.png)

I had to run `npm update fastmatter` in order for it to work.

yamgent commented on docs/userGuide/components/searchbar.md:
```diff
@@ -40,3 +40,5 @@ placeholder | `String` | `''` | The placeholder text shown when no keywords are
 
 %%{{ icon_info }} Related topic: [User Guide: Making the Site Searchable]({{ baseUrl }}/userGuide/makingTheSiteSearchable.html).%%
+
+If you do not wish to use MarkBind's searchbar, it may be helpful to include the option `disableSearch: true` in your `site.json`. This stops MarkBind from indexing search headings and speeds up building.
```
Seems like this can be moved to `Making the Site Searchable` and be elaborated a bit further (that the wish to not use MarkBind's searching functionality, may be because they already have an external service provider).

yamgent commented on docs/userGuide/siteConfiguration.md:
```diff
@@ -118,4 +118,8 @@ The ignore pattern follows the [glob pattern used in .gitignore](https://git-scm
 
 **The level of headings to be indexed for searching.** Default: `3` %%i.e., only headings of levels 1,2,3 will be indexed for searching%%.
 
+#### **`disableSearch`**
```
I think using a positive term `enableSearch` would be better (this is similar to `searchable` for pages).

yamgent commented on docs/userGuide/siteConfiguration.md:
```diff
@@ -118,4 +118,8 @@ The ignore pattern follows the [glob pattern used in .gitignore](https://git-scm
 
+#### **`disableSearch`**
+
+**If set to `true`, MarkBind will not index pages for searching.** See [Use Searchbar](usingComponents.html#search-bar) for more information.
```
```
**Specifies that the website should use MarkBind's search functionality.** See [User Guide: Making the Site Searchable](userguide/makingthesitesearchable) for more details.
```

jamos-tay replied:
Should this be true by default?

yamgent replied:
Yes it should be true by default for `enableSearch`.

yamgent commented on src/Site.js:
```diff
@@ -306,7 +306,7 @@ Site.prototype.createPage = function (config) {
     pageTemplate: this.pageTemplate,
     rootPath: this.rootPath,
-    searchable: config.searchable,
+    searchable: (!('enableSearch' in this.siteConfig) || this.siteConfig.enableSearch) && config.searchable,
```
Possible to just set the variable during `readSiteConfig()`, so that it is simplified to `searchable: this.siteConfig.enableSearch && config.searchable`?

```diff
diff --git a/src/Site.js b/src/Site.js
index 3373f29..2e41cb2 100644
--- a/src/Site.js
+++ b/src/Site.js
@@ -274,6 +274,9 @@ Site.prototype.readSiteConfig = function (baseUrl) {
       .then((config) => {
         this.siteConfig = config;
         this.siteConfig.baseUrl = (baseUrl === undefined) ? this.siteConfig.baseUrl : baseUrl;
+        if (this.siteConfig.enableSearch === undefined) {
+          this.siteConfig.enableSearch = true;
+        }
         resolve(this.siteConfig);
       })
       .catch((err) => {
```

yamgent commented on src/Site.js:
```diff
@@ -700,6 +700,7 @@ Site.prototype.writeSiteData = function () {
   return new Promise((resolve, reject) => {
     const siteDataPath = path.join(this.outputPath, SITE_DATA_NAME);
     const siteData = {
+      enableSearch: !('enableSearch' in this.siteConfig) || this.siteConfig.enableSearch,
```
Same as above. It just becomes `enableSearch: this.siteConfig.enableSearch`.

jamos-tay commented:
Updated

acjh commented on src/Site.js:
```diff
@@ -274,6 +274,9 @@ Site.prototype.readSiteConfig = function (baseUrl) {
         this.siteConfig.baseUrl = (baseUrl === undefined) ? this.siteConfig.baseUrl : baseUrl;
+        if (this.siteConfig.enableSearch === undefined) {
+          this.siteConfig.enableSearch = true;
+        }
```
we can do this in one line:
```js
this.siteConfig.enableSearch = (config.enableSearch === undefined) || config.enableSearch;
```

acjh commented on test/unit/Site.test.js:
```diff
@@ -203,10 +203,11 @@ test('Site read site config for default', async () => {
     'site.json': SITE_JSON_DEFAULT,
   };
   fs.vol.fromJSON(json, '');
+  const expectedSiteJson = { ...JSON.parse(SITE_JSON_DEFAULT), ...{ enableSearch: true } };
```
how about:
```diff
  fs.vol.fromJSON(json, '');
- const expectedSiteJson = { ...JSON.parse(SITE_JSON_DEFAULT), ...{ enableSearch: true } };

+ const siteConfigDefaults = { enableSearch: true };
+ const expectedSiteConfig = { ...JSON.parse(SITE_JSON_DEFAULT), ...siteConfigDefaults };
  const site = new Site('./', '_site');
  await site.readSiteConfig();
- expect(site.siteConfig).toEqual(expectedSiteJson);
+ expect(site.siteConfig).toEqual(expectedSiteConfig);
```

yamgent commented:
@jamos-tay hey, when you are back to working on your PRs, is it possible for you to give the utmost priority to this PR? This is because Prof wants to use an external search provider for the upcoming semester CS2103T's website, and this feature is one of the critical ones that we need in order to make it happen. Thanks! :)

jamos-tay commented:
Sure no problem, updated

yamgent commented:
I tried integrating the external search provider on `se-book` and used your feature, but it still doesn't seem to solve the problems that we had. Will make a further investigation about this. :S

Thanks for updating the PR anyway!

yamgent commented on docs/userGuide/siteConfiguration.md:
```diff
@@ -118,4 +118,8 @@ The ignore pattern follows the [glob pattern used in .gitignore](https://git-scm
 
+#### **`enableSearch`**
+
+**Specifies that the website should use MarkBind's search functionality.** Default: `true`. See [User Guide: Making the Site Searchable](userguide/makingthesitesearchable) for more details.
```
Link is broken, should be `[User Guide: Making the Site Searchable]({{ baseUrl }}/userGuide/makingTheSiteSearchable.html)`.

jamos-tay commented:
Updated
